File: composeApp/src/androidMain/kotlin/com/kissangram/repository/FirestoreStoryRepository.kt
package com.kissangram.repository

import android.util.Log
import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import com.kissangram.model.*
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import com.google.firebase.firestore.Query
import java.util.Date

/**
 * Firestore implementation of StoryRepository.
 * Handles story creation and retrieval per FIRESTORE_SCHEMA.md.
 */
class FirestoreStoryRepository(
    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance(
        com.google.firebase.FirebaseApp.getInstance(),
        DATABASE_NAME
    ),
    private val authRepository: AuthRepository
) : StoryRepository {
    
    private val storiesCollection
        get() = firestore.collection("stories")
    
    override suspend fun createStory(storyData: Map<String, Any?>): Story {
        Log.d(TAG, "📤 FirestoreStoryRepository: Starting story creation in Firestore")
        Log.d(TAG, "📤 FirestoreStoryRepository: Story Data Summary:")
        Log.d(TAG, "   - Author ID: ${storyData["authorId"]}")
        Log.d(TAG, "   - Author Name: ${storyData["authorName"]}")
        Log.d(TAG, "   - Visibility: ${storyData["visibility"]}")
        Log.d(TAG, "   - Media URL: ${(storyData["media"] as? Map<*, *>)?.get("url")}")
        Log.d(TAG, "   - Location: ${(storyData["location"] as? Map<*, *>)?.get("name") ?: "none"}")
        Log.d(TAG, "   - Text Overlay: ${if (storyData["textOverlay"] != null) "yes" else "no"}")
        
        // Let Firestore generate the document ID automatically
        val storyRef = storiesCollection.document()
        val storyId = storyRef.id
        Log.d(TAG, "📤 FirestoreStoryRepository: Generated Story ID: $storyId")
        
        // Build Firestore document data
        val documentData = mutableMapOf<String, Any?>()
        documentData.putAll(storyData)
        documentData["id"] = storyId  // Add the generated ID to the data
        
        // Add server timestamp for createdAt
        documentData["createdAt"] = com.google.firebase.firestore.FieldValue.serverTimestamp()
        
        // Calculate expiresAt as createdAt + 24 hours
        // We'll set this after getting the actual createdAt timestamp
        // For now, we'll use a client-side estimate (Cloud Function will set it properly later)
        val estimatedExpiresAt = Date(System.currentTimeMillis() + (24 * 60 * 60 * 1000L))
        documentData["expiresAt"] = Timestamp(estimatedExpiresAt)
        
        return try {
            // Create story document
            Log.d(TAG, "📤 FirestoreStoryRepository: Sending document to Firestore...")
            storyRef.set(documentData).await()
            
            Log.d(TAG, "✅ FirestoreStoryRepository: Story created successfully in Firestore")
            Log.d(TAG, "   - Story ID: $storyId")
            
            // Fetch the created story to return it
            val doc = storyRef.get().await()
            doc.toStory() ?: throw Exception("Failed to retrieve created story")
        } catch (e: Exception) {
            Log.e(TAG, "createStory: FAILED", e)
            throw Exception("Failed to create story: ${e.message}", e)
        }
    }
    
    override suspend fun getStoryBar(): List<UserStories> {
        Log.d(TAG, "getStoryBar: start")
        val currentUserId = authRepository.getCurrentUserId()
        if (currentUserId == null) {
            Log.w(TAG, "getStoryBar: currentUserId is null, returning empty list")
            return emptyList()
        }
        Log.d(TAG, "getStoryBar: currentUserId=${currentUserId.take(8)}...")
        
        return try {
            // Query storiesFeed subcollection ordered by createdAt DESC
            val storiesFeedRef = firestore
                .collection("users")
                .document(currentUserId)
                .collection("storiesFeed")
                .orderBy("createdAt", Query.Direction.DESCENDING)
            
            val snapshot = storiesFeedRef.get().await()
            val rawCount = snapshot.documents.size
            Log.d(TAG, "getStoryBar: fetched $rawCount story documents")
            
            if (rawCount == 0) {
                return emptyList()
            }
            
            val currentTime = System.currentTimeMillis()
            
            // Filter expired stories and parse to Story objects
            val stories = snapshot.documents
                .mapNotNull { doc ->
                    try {
                        // Check if story is expired
                        val expiresAt = doc.getTimestamp("expiresAt")?.toDate()?.time
                        if (expiresAt != null && expiresAt < currentTime) {
                            Log.d(TAG, "getStoryBar: skipping expired story ${doc.id}")
                            return@mapNotNull null
                        }
                        
                        doc.toStory()
                    } catch (e: Exception) {
                        Log.e(TAG, "getStoryBar: failed to parse story ${doc.id}", e)
                        null
                    }
                }
            
            Log.d(TAG, "getStoryBar: parsed ${stories.size} valid stories (after filtering expired)")
            
            if (stories.isEmpty()) {
                return emptyList()
            }
            
            // Batch check views and likes
            val storyIds = stories.map { it.id }
            val viewedStoryIds = batchCheckViews(storyIds, currentUserId)
            val likedStoryIds = batchCheckLikes(storyIds, currentUserId)
            
            // Update stories with view/like status
            val storiesWithStatus = stories.map { story ->
                story.copy(
                    isViewedByMe = viewedStoryIds.contains(story.id),
                    isLikedByMe = likedStoryIds.contains(story.id)
                )
            }
            
            // Group stories by authorId
            val storiesByAuthor = storiesWithStatus.groupBy { it.authorId }
            Log.d(TAG, "getStoryBar: grouped into ${storiesByAuthor.size} authors")
            
            // Create UserStories objects
            val userStoriesList = storiesByAuthor.map { (authorId, authorStories) ->
                val firstStory = authorStories.first()
                val sortedStories = authorStories.sortedByDescending { it.createdAt }
                val hasUnviewedStories = authorStories.any { !it.isViewedByMe }
                val latestStoryTime = authorStories.maxOfOrNull { it.createdAt } ?: 0L
                
                UserStories(
                    userId = authorId,
                    userName = firstStory.authorName,
                    userProfileImageUrl = firstStory.authorProfileImageUrl,
                    userRole = firstStory.authorRole,
                    userVerificationStatus = firstStory.authorVerificationStatus,
                    stories = sortedStories,
                    hasUnviewedStories = hasUnviewedStories,
                    latestStoryTime = latestStoryTime
                )
            }
            
            // Sort by latestStoryTime DESC
            val sortedUserStories = userStoriesList.sortedByDescending { it.latestStoryTime }
            
            Log.d(TAG, "getStoryBar: returning ${sortedUserStories.size} user stories")
            sortedUserStories
            
        } catch (e: Exception) {
            Log.e(TAG, "getStoryBar: failed", e)
            emptyList() // Return empty list on error
        }
    }
    
    /**
     * Batch check which stories are viewed by the current user.
     */
    private suspend fun batchCheckViews(storyIds: List<String>, userId: String): Set<String> = coroutineScope {
        if (storyIds.isEmpty()) return@coroutineScope emptySet()
        
        try {
            val results = storyIds.map { storyId ->
                async {
                    try {
                        val viewDoc = firestore
                            .collection("stories")
                            .document(storyId)
                            .collection("views")
                            .document(userId)
                            .get()
                            .await()
                        if (viewDoc.exists()) storyId else null
                    } catch (e: Exception) {
                        Log.e(TAG, "batchCheckViews: failed for storyId=$storyId", e)
                        null
                    }
                }
            }
            
            val viewedStoryIds = results.mapNotNull { it.await() }.toSet()
            Log.d(TAG, "batchCheckViews: checked ${storyIds.size} stories, found ${viewedStoryIds.size} viewed")
            viewedStoryIds
        } catch (e: Exception) {
            Log.e(TAG, "batchCheckViews: failed", e)
            emptySet()
        }
    }
    
    /**
     * Batch check which stories are liked by the current user.
     */
    private suspend fun batchCheckLikes(storyIds: List<String>, userId: String): Set<String> = coroutineScope {
        if (storyIds.isEmpty()) return@coroutineScope emptySet()
        
        try {
            val results = storyIds.map { storyId ->
                async {
                    try {
                        val likeDoc = firestore
                            .collection("stories")
                            .document(storyId)
                            .collection("likes")
                            .document(userId)
                            .get()
                            .await()
                        if (likeDoc.exists()) storyId else null
                    } catch (e: Exception) {
                        Log.e(TAG, "batchCheckLikes: failed for storyId=$storyId", e)
                        null
                    }
                }
            }
            
            val likedStoryIds = results.mapNotNull { it.await() }.toSet()
            Log.d(TAG, "batchCheckLikes: checked ${storyIds.size} stories, found ${likedStoryIds.size} liked")
            likedStoryIds
        } catch (e: Exception) {
            Log.e(TAG, "batchCheckLikes: failed", e)
            emptySet()
        }
    }
    
    override suspend fun getStoriesForUser(userId: String): List<Story> {
        // TODO: Implement get stories for user
        throw UnsupportedOperationException("Not yet implemented")
    }
    
    override suspend fun markStoryAsViewed(storyId: String) {
        val currentUserId = authRepository.getCurrentUserId()
        if (currentUserId == null) {
            Log.w(TAG, "markStoryAsViewed: currentUserId is null")
            return
        }
        
        try {
            // Check if already viewed to avoid duplicate writes
            val viewDocRef = firestore
                .collection("stories")
                .document(storyId)
                .collection("views")
                .document(currentUserId)
            
            val existingView = viewDocRef.get().await()
            if (existingView.exists()) {
                Log.d(TAG, "markStoryAsViewed: story $storyId already viewed by user")
                return
            }
            
            // Create view document
            val viewData = mapOf(
                "id" to currentUserId,
                "viewedAt" to com.google.firebase.firestore.FieldValue.serverTimestamp()
            )
            
            viewDocRef.set(viewData).await()
            
            // Increment viewsCount on the story document
            firestore
                .collection("stories")
                .document(storyId)
                .update("viewsCount", com.google.firebase.firestore.FieldValue.increment(1))
                .await()
            
            Log.d(TAG, "markStoryAsViewed: marked story $storyId as viewed")
        } catch (e: Exception) {
            Log.e(TAG, "markStoryAsViewed: failed for storyId=$storyId", e)
            throw Exception("Failed to mark story as viewed: ${e.message}", e)
        }
    }
    
    override suspend fun getMyStories(): List<Story> {
        // TODO: Implement get my stories
        throw UnsupportedOperationException("Not yet implemented")
    }
    
    /**
     * Extension function to convert Firestore DocumentSnapshot to Story model
     */
    private fun com.google.firebase.firestore.DocumentSnapshot.toStory(): Story? {
        return try {
            val id = getString("id") ?: this.id
            val authorId = getString("authorId") ?: return null
            val authorName = getString("authorName") ?: return null
            val authorUsername = getString("authorUsername") ?: return null
            val authorProfileImageUrl = getString("authorProfileImageUrl")
            val authorRoleStr = getString("authorRole") ?: "farmer"
            val authorVerificationStatusStr = getString("authorVerificationStatus") ?: "unverified"
            
            // Parse media
            val mediaMap = get("media") as? Map<String, Any> ?: return null
            val mediaUrl = mediaMap["url"] as? String ?: return null
            val mediaTypeStr = mediaMap["type"] as? String ?: "image"
            val mediaType = when (mediaTypeStr) {
                "video" -> MediaType.VIDEO
                else -> MediaType.IMAGE
            }
            val thumbnailUrl = mediaMap["thumbnailUrl"] as? String
            
            val storyMedia = StoryMedia(
                url = mediaUrl,
                type = mediaType,
                thumbnailUrl = thumbnailUrl
            )
            
            // Parse text overlay (optional)
            val textOverlayMap = get("textOverlay") as? Map<String, Any>
            val textOverlay = textOverlayMap?.let {
                val text = it["text"] as? String ?: return@let null
                val positionMap = it["position"] as? Map<String, Any>
                val positionX = (positionMap?.get("x") as? Number)?.toFloat() ?: return@let null
                val positionY = (positionMap?.get("y") as? Number)?.toFloat() ?: return@let null
                TextOverlay(
                    text = text,
                    positionX = positionX,
                    positionY = positionY
                )
            }
            
            // Parse location (optional) - per schema: location: { name: "..." }
            val locationMap = get("location") as? Map<String, Any>
            val locationName = locationMap?.get("name") as? String
            
            // Parse visibility
            val visibilityStr = getString("visibility") ?: "public"
            val visibility = when (visibilityStr) {
                "followers" -> PostVisibility.FOLLOWERS
                else -> PostVisibility.PUBLIC
            }
            
            val viewsCount = (getLong("viewsCount") ?: 0L).toInt()
            val likesCount = (getLong("likesCount") ?: 0L).toInt()
            
            val createdAt = getTimestamp("createdAt")?.toDate()?.time ?: System.currentTimeMillis()
            val expiresAt = getTimestamp("expiresAt")?.toDate()?.time ?: (createdAt + 24 * 60 * 60 * 1000L)
            
            Story(
                id = id,
                authorId = authorId,
                authorName = authorName,
                authorUsername = authorUsername,
                authorProfileImageUrl = authorProfileImageUrl,
                authorRole = stringToUserRole(authorRoleStr),
                authorVerificationStatus = stringToVerificationStatus(authorVerificationStatusStr),
                media = storyMedia,
                textOverlay = textOverlay,
                locationName = locationName,
                visibility = visibility,
                viewsCount = viewsCount,
                likesCount = likesCount,
                isViewedByMe = false, // TODO: Check from views subcollection
                isLikedByMe = false, // TODO: Check from likes subcollection
                createdAt = createdAt,
                expiresAt = expiresAt
            )
        } catch (e: Exception) {
            Log.e(TAG, "toStory: Failed to parse story document", e)
            null
        }
    }
    
    private fun stringToUserRole(roleStr: String): UserRole {
        return when (roleStr) {
            "expert" -> UserRole.EXPERT
            "agripreneur" -> UserRole.AGRIPRENEUR
            "input_seller" -> UserRole.INPUT_SELLER
            "agri_lover" -> UserRole.AGRI_LOVER
            else -> UserRole.FARMER
        }
    }
    
    private fun stringToVerificationStatus(statusStr: String): VerificationStatus {
        return when (statusStr) {
            "pending" -> VerificationStatus.PENDING
            "verified" -> VerificationStatus.VERIFIED
            "rejected" -> VerificationStatus.REJECTED
            else -> VerificationStatus.UNVERIFIED
        }
    }
    
    companion object {
        private const val TAG = "FirestoreStoryRepo"
        private const val DATABASE_NAME = "kissangram"
    }
}

File: iosApp/iosApp/repository/FirestoreStoryRepository.swift
import Foundation
import os.log
import FirebaseFirestore
import Shared

/// Firestore implementation of StoryRepository.
/// Handles story creation and retrieval per FIRESTORE_SCHEMA.md.
final class FirestoreStoryRepository: StoryRepository {
    private static let log = Logger(subsystem: "com.kissangram", category: "FirestoreStoryRepo")
    private let firestore = Firestore.firestore(database: "kissangram")
    private let authRepository: AuthRepository
    private let followRepository: FollowRepository
    
    init(authRepository: AuthRepository, followRepository: FollowRepository) {
        self.authRepository = authRepository
        self.followRepository = followRepository
    }
    
    private var storiesCollection: CollectionReference {
        firestore.collection(Self.collectionStories)
    }
    func createStory(storyData: [String: Any]) async throws -> Story {
        Self.log.debug("createStory: Starting story creation in Firestore")
        Self.log.debug("createStory: Story Data Summary:")
     
        let storyRef = storiesCollection.document()
        let storyId = storyRef.documentID
        Self.log.debug("createStory: Generated Story ID: \(storyId)")

        // Build Firestore document data
        var documentData: [String: Any] = [:]

        for (key, value) in storyData {
            documentData[key] = value
        }

        // Add generated ID
        documentData["id"] = storyId

        // Server timestamp
        documentData["createdAt"] = FieldValue.serverTimestamp()

        // expiresAt = now + 24h (client-side estimate)
        let estimatedExpiresAt = Date().addingTimeInterval(24 * 60 * 60)
        documentData["expiresAt"] = Timestamp(date: estimatedExpiresAt)

        do {
            // Create document
            Self.log.debug("createStory: Sending document to Firestore...")
            try await storyRef.setData(documentData)

            Self.log.debug("createStory: Story created successfully in Firestore")
            Self.log.debug("   - Story ID: \(storyId)")

            // Fetch created document
            let doc = try await storyRef.getDocument()

            guard let story = toStory(from: doc) else {
                Self.log.error("createStory: Failed to retrieve created story")
                throw NSError(
                    domain: "FirestoreStoryRepository",
                    code: 500,
                    userInfo: [NSLocalizedDescriptionKey: "Failed to retrieve created story"]
                )
            }
            
            return story
        } catch {
            Self.log.error("createStory: FAILED - \(error.localizedDescription)")
            throw error
        }
    }
    

    
   
   
    
 
    
    
    
    func getStoryBar() async throws -> [UserStories] {
        Self.log.debug("getStoryBar: start")
        guard let currentUserId = try await authRepository.getCurrentUserId() else {
            Self.log.warning("getStoryBar: currentUserId is nil, returning empty")
            return []
        }
        let uidPrefix = String(currentUserId.prefix(8))
        Self.log.debug("getStoryBar: currentUserId=\(uidPrefix)...")
        
        do {
            // Get list of followed user IDs from users/{currentUserId}/following
            let followingSnapshot = try await firestore
                .collection("users")
                .document(currentUserId)
                .collection("following")
                .getDocuments()
            
            var followedUserIds = Set(followingSnapshot.documents.map { $0.documentID })
            // Always include current user's own ID
            followedUserIds.insert(currentUserId)
            
            Self.log.debug("getStoryBar: found \(followedUserIds.count) users to query (including self)")
            
            if followedUserIds.isEmpty {
                Self.log.debug("getStoryBar: no followed users, returning empty")
                return []
            }
            
            // Query all active stories from /stories collection
            // Filter by followed users client-side to avoid whereIn limit of 10
            let storiesSnapshot = try await firestore
                .collection("stories")
                .whereField("isActive", isEqualTo: true)
                .order(by: "createdAt", descending: true)
                .getDocuments()
            
            let rawCount = storiesSnapshot.documents.count
            Self.log.debug("getStoryBar: fetched \(rawCount) active story documents")
            
            if rawCount == 0 {
                return []
            }
            
            let currentTime = Int64(Date().timeIntervalSince1970 * 1000)
            
            // Filter by followed users, expired stories, and parse to Story objects
            let stories = storiesSnapshot.documents.compactMap { doc -> Story? in
                do {
                    guard let authorId = doc.get("authorId") as? String,
                          followedUserIds.contains(authorId) else {
                        return nil
                    }
                    
                    // Check if story is expired
                    if let expiresAtTimestamp = doc.get("expiresAt") as? Timestamp {
                        let expiresAt = Int64(expiresAtTimestamp.dateValue().timeIntervalSince1970 * 1000)
                        if expiresAt < currentTime {
                            Self.log.debug("getStoryBar: skipping expired story \(doc.documentID)")
                            return nil
                        }
                    }
                    
                    return toStory(from: doc)
                } catch {
                    Self.log.error("getStoryBar: failed to parse story \(doc.documentID): \(error.localizedDescription)")
                    return nil
                }
            }
            
            Self.log.debug("getStoryBar: parsed \(stories.count) valid stories (after filtering by followed users and expired)")
            
            if stories.isEmpty {
                return []
            }
            
            // Batch check views and likes
            let storyIds = stories.map { $0.id }
            let viewedStoryIds = try await batchCheckViews(storyIds: storyIds, userId: currentUserId)
            let likedStoryIds = try await batchCheckLikes(storyIds: storyIds, userId: currentUserId)
            
            // Update stories with view/like status
            let storiesWithStatus = stories.map { story in
                Story(
                    id: story.id,
                    authorId: story.authorId,
                    authorName: story.authorName,
                    authorUsername: story.authorUsername,
                    authorProfileImageUrl: story.authorProfileImageUrl,
                    authorRole: story.authorRole,
                    authorVerificationStatus: story.authorVerificationStatus,
                    media: story.media,
                    textOverlay: story.textOverlay,
                    locationName: story.locationName,
                    visibility: story.visibility,
                    viewsCount: story.viewsCount,
                    likesCount: story.likesCount,
                    isViewedByMe: viewedStoryIds.contains(story.id),
                    isLikedByMe: likedStoryIds.contains(story.id),
                    createdAt: story.createdAt,
                    expiresAt: story.expiresAt
                )
            }
            
            // Group stories by authorId
            let storiesByAuthor = Dictionary(grouping: storiesWithStatus, by: { $0.authorId })
            Self.log.debug("getStoryBar: grouped into \(storiesByAuthor.count) authors")
            
            // Create UserStories objects
            let userStoriesList = storiesByAuthor.map { (authorId, authorStories) in
                let firstStory = authorStories.first!
                let sortedStories = authorStories.sorted { $0.createdAt > $1.createdAt }
                let hasUnviewedStories = authorStories.contains { !$0.isViewedByMe }
                let latestStoryTime = authorStories.map { $0.createdAt }.max() ?? 0
                
                return UserStories(
                    userId: authorId,
                    userName: firstStory.authorName,
                    userProfileImageUrl: firstStory.authorProfileImageUrl,
                    userRole: firstStory.authorRole,
                    userVerificationStatus: firstStory.authorVerificationStatus,
                    stories: sortedStories,
                    hasUnviewedStories: hasUnviewedStories,
                    latestStoryTime: latestStoryTime
                )
            }
            
            // Sort by latestStoryTime DESC
            let sortedUserStories = userStoriesList.sorted { $0.latestStoryTime > $1.latestStoryTime }
            
            Self.log.debug("getStoryBar: returning \(sortedUserStories.count) user stories")
            return sortedUserStories
            
        } catch {
            Self.log.error("getStoryBar: failed: \(error.localizedDescription)")
            return [] // Return empty array on error
        }
    }
    
    /// Batch check which stories are viewed by the current user.
    private func batchCheckViews(storyIds: [String], userId: String) async throws -> Set<String> {
        if storyIds.isEmpty { return Set<String>() }
        
        return try await withThrowingTaskGroup(of: (String, Bool).self) { group in
            var viewedStoryIds = Set<String>()
            
            // Create tasks to check each story
            for storyId in storyIds {
                group.addTask {
                    let viewDoc = try await self.firestore
                        .collection("stories")
                        .document(storyId)
                        .collection("views")
                        .document(userId)
                        .getDocument()
                    return (storyId, viewDoc.exists)
                }
            }
            
            // Collect results
            for try await (storyId, exists) in group {
                if exists {
                    viewedStoryIds.insert(storyId)
                }
            }
            
            Self.log.debug("batchCheckViews: checked \(storyIds.count) stories, found \(viewedStoryIds.count) viewed")
            return viewedStoryIds
        }
    }
    
    /// Batch check which stories are liked by the current user.
    private func batchCheckLikes(storyIds: [String], userId: String) async throws -> Set<String> {
        if storyIds.isEmpty { return Set<String>() }
        
        return try await withThrowingTaskGroup(of: (String, Bool).self) { group in
            var likedStoryIds = Set<String>()
            
            // Create tasks to check each story
            for storyId in storyIds {
                group.addTask {
                    let likeDoc = try await self.firestore
                        .collection("stories")
                        .document(storyId)
                        .collection("likes")
                        .document(userId)
                        .getDocument()
                    return (storyId, likeDoc.exists)
                }
            }
            
            // Collect results
            for try await (storyId, exists) in group {
                if exists {
                    likedStoryIds.insert(storyId)
                }
            }
            
            Self.log.debug("batchCheckLikes: checked \(storyIds.count) stories, found \(likedStoryIds.count) liked")
            return likedStoryIds
        }
    }

    func getStoriesForUser(userId: String) async throws -> [Story] {
        Self.log.debug("getStoriesForUser: start for userId=\(String(userId.prefix(8)))...")
        guard let currentUserId = try await authRepository.getCurrentUserId() else {
            Self.log.warning("getStoriesForUser: currentUserId is nil, returning empty")
            return []
        }
        
        do {
            // Check if current user follows the target user
            let isFollowing: Bool
            if currentUserId == userId {
                isFollowing = true // User viewing their own profile
            } else {
                do {
                    let isFollowingResult = try await followRepository.isFollowing(userId: userId)
                    isFollowing = isFollowingResult.boolValue
                } catch {
                    Self.log.error("getStoriesForUser: failed to check follow status: \(error.localizedDescription)")
                    isFollowing = false
                }
            }
            
            Self.log.debug("getStoriesForUser: isFollowing=\(isFollowing)")
            
            // Build query: authorId == userId, isActive == true
            var query = firestore
                .collection("stories")
                .whereField("authorId", isEqualTo: userId)
                .whereField("isActive", isEqualTo: true)
                .order(by: "createdAt", descending: true)
            
            // If not following, only show public stories
            if !isFollowing {
                query = query.whereField("visibility", isEqualTo: "public")
            }
            // If following, show all stories (both public and followers)
            
            let snapshot = try await query.getDocuments()
            let rawCount = snapshot.documents.count
            Self.log.debug("getStoriesForUser: fetched \(rawCount) story documents")
            
            if rawCount == 0 {
                return []
            }
            
            let currentTime = Int64(Date().timeIntervalSince1970 * 1000)
            
            // Filter expired stories and parse to Story objects
            let stories = snapshot.documents.compactMap { doc -> Story? in
                do {
                    // Check if story is expired
                    if let expiresAtTimestamp = doc.get("expiresAt") as? Timestamp {
                        let expiresAt = Int64(expiresAtTimestamp.dateValue().timeIntervalSince1970 * 1000)
                        if expiresAt < currentTime {
                            Self.log.debug("getStoriesForUser: skipping expired story \(doc.documentID)")
                            return nil
                        }
                    }
                    
                    return toStory(from: doc)
                } catch {
                    Self.log.error("getStoriesForUser: failed to parse story \(doc.documentID): \(error.localizedDescription)")
                    return nil
                }
            }
            
            Self.log.debug("getStoriesForUser: parsed \(stories.count) valid stories (after filtering expired)")
            
            if stories.isEmpty {
                return []
            }
            
            // Batch check views and likes
            let storyIds = stories.map { $0.id }
            let viewedStoryIds = try await batchCheckViews(storyIds: storyIds, userId: currentUserId)
            let likedStoryIds = try await batchCheckLikes(storyIds: storyIds, userId: currentUserId)
            
            // Update stories with view/like status
            let storiesWithStatus = stories.map { story in
                Story(
                    id: story.id,
                    authorId: story.authorId,
                    authorName: story.authorName,
                    authorUsername: story.authorUsername,
                    authorProfileImageUrl: story.authorProfileImageUrl,
                    authorRole: story.authorRole,
                    authorVerificationStatus: story.authorVerificationStatus,
                    media: story.media,
                    textOverlay: story.textOverlay,
                    locationName: story.locationName,
                    visibility: story.visibility,
                    viewsCount: story.viewsCount,
                    likesCount: story.likesCount,
                    isViewedByMe: viewedStoryIds.contains(story.id),
                    isLikedByMe: likedStoryIds.contains(story.id),
                    createdAt: story.createdAt,
                    expiresAt: story.expiresAt
                )
            }
            
            Self.log.debug("getStoriesForUser: returning \(storiesWithStatus.count) stories")
            return storiesWithStatus
            
        } catch {
            Self.log.error("getStoriesForUser: failed: \(error.localizedDescription)")
            throw error
        }
    }

    func getMyStories() async throws -> [Story] {
        guard let currentUserId = try await authRepository.getCurrentUserId() else {
            Self.log.warning("getMyStories: currentUserId is nil, returning empty")
            return []
        }
        // Use getStoriesForUser with current user's ID
        return try await getStoriesForUser(userId: currentUserId)
    }
    func markStoryAsViewed(storyId: String) async throws {
        guard let currentUserId = try await authRepository.getCurrentUserId() else {
            Self.log.warning("markStoryAsViewed: currentUserId is nil")
            return
        }
        
        do {
            // Check if already viewed to avoid duplicate writes
            let viewDocRef = firestore
                .collection("stories")
                .document(storyId)
                .collection("views")
                .document(currentUserId)
            
            let existingView = try await viewDocRef.getDocument()
            if existingView.exists {
                Self.log.debug("markStoryAsViewed: story \(storyId) already viewed by user")
                return
            }
            
            // Create view document
            let viewData: [String: Any] = [
                "id": currentUserId,
                "viewedAt": FieldValue.serverTimestamp()
            ]
            
            try await viewDocRef.setData(viewData)
            
            // Increment viewsCount on the story document
            try await firestore
                .collection("stories")
                .document(storyId)
                .updateData(["viewsCount": FieldValue.increment(Int64(1))])
            
            Self.log.debug("markStoryAsViewed: marked story \(storyId) as viewed")
        } catch {
            Self.log.error("markStoryAsViewed: failed for storyId=\(storyId): \(error.localizedDescription)")
            throw error
        }
    }
    
  
    
    // MARK: - Helper Methods
    
    private func toStory(from doc: DocumentSnapshot) -> Story? {
        guard let data = doc.data() else { return nil }
        
        do {
            let id = (doc.get("id") as? String) ?? doc.documentID
            let authorId = (doc.get("authorId") as? String) ?? ""
            let authorName = (doc.get("authorName") as? String) ?? ""
            let authorUsername = (doc.get("authorUsername") as? String) ?? ""
            let authorProfileImageUrl = doc.get("authorProfileImageUrl") as? String
            let authorRoleStr = (doc.get("authorRole") as? String) ?? "farmer"
            let authorVerificationStatusStr = (doc.get("authorVerificationStatus") as? String) ?? "unverified"
            
            // Parse media
            guard let mediaMap = data["media"] as? [String: Any],
                  let mediaUrl = mediaMap["url"] as? String else {
                return nil
            }
            let mediaTypeStr = mediaMap["type"] as? String ?? "image"
            let mediaType: Shared.MediaType = mediaTypeStr == "video" ? .video : .image
            let thumbnailUrl = mediaMap["thumbnailUrl"] as? String
            
            let storyMedia = StoryMedia(
                url: mediaUrl,
                type: mediaType,
                thumbnailUrl: thumbnailUrl?.isEmpty == false ? thumbnailUrl : nil
            )
            
            // Parse text overlay (optional) - per schema: textOverlay: { text: "...", position: { x, y } }
            let textOverlayMap = data["textOverlay"] as? [String: Any]
            let textOverlay: TextOverlay? = textOverlayMap.flatMap { map in
                guard let text = map["text"] as? String,
                      let positionMap = map["position"] as? [String: Any],
                      let positionX = positionMap["x"] as? NSNumber,
                      let positionY = positionMap["y"] as? NSNumber else {
                    return nil
                }
                return TextOverlay(
                    text: text,
                    positionX: positionX.floatValue,
                    positionY: positionY.floatValue
                )
            }
            
            // Parse location (optional) - per schema: location: { name: "..." }
            let locationMap = data["location"] as? [String: Any]
            let locationName = locationMap?["name"] as? String
            
            // Parse visibility
            let visibilityStr = (doc.get("visibility") as? String) ?? "public"
            let visibility: Shared.PostVisibility = visibilityStr == "followers" ? .followers : .public_
            
            let viewsCount = (data["viewsCount"] as? Int) ?? 0
            let likesCount = (data["likesCount"] as? Int) ?? 0
            
            // Parse timestamps
            let createdAt: Int64
            if let createdAtTimestamp = doc.get("createdAt") as? Timestamp {
                createdAt = Int64(createdAtTimestamp.dateValue().timeIntervalSince1970 * 1000)
            } else if let created = data["createdAt"] as? Int64 {
                createdAt = created
            } else {
                createdAt = Int64(Date().timeIntervalSince1970 * 1000)
            }
            
            let expiresAt: Int64
            if let expiresAtTimestamp = doc.get("expiresAt") as? Timestamp {
                expiresAt = Int64(expiresAtTimestamp.dateValue().timeIntervalSince1970 * 1000)
            } else if let expired = data["expiresAt"] as? Int64 {
                expiresAt = expired
            } else {
                // Default to 24 hours after createdAt
                expiresAt = createdAt + (24 * 60 * 60 * 1000)
            }
            
            return Story(
                id: id,
                authorId: authorId,
                authorName: authorName,
                authorUsername: authorUsername,
                authorProfileImageUrl: authorProfileImageUrl,
                authorRole: stringToUserRole(authorRoleStr),
                authorVerificationStatus: stringToVerificationStatus(authorVerificationStatusStr),
                media: storyMedia,
                textOverlay: textOverlay,
                locationName: locationName,
                visibility: visibility,
                viewsCount: Int32(viewsCount),
                likesCount: Int32(likesCount),
                isViewedByMe: false, // TODO: Check from views subcollection
                isLikedByMe: false, // TODO: Check from likes subcollection
                createdAt: Int64(createdAt),
                expiresAt: Int64(expiresAt)
            )
        } catch {
            return nil
        }
    }
    
    private func stringToUserRole(_ roleStr: String) -> UserRole {
        switch roleStr {
        case "expert": return .expert
        case "agripreneur": return .agripreneur
        case "input_seller": return .inputSeller
        case "agri_lover": return .agriLover
        default: return .farmer
        }
    }
    
    private func stringToVerificationStatus(_ statusStr: String) -> VerificationStatus {
        switch statusStr {
        case "pending": return .pending
        case "verified": return .verified
        case "rejected": return .rejected
        default: return .unverified
        }
    }
    
    private static let collectionStories = "stories"
}
